lib/pos_blockchain.go

```diff
@@ -0,0 +1,1134 @@
+	utxoView, err := bc.getUtxoViewAtBlockHash(*block.Header.PrevBlockHash)
+	if err != nil {
+		return errors.Wrapf(err, "commitBlockPoS: Problem initializing UtxoView: ")
+	}
+	txHashes := collections.Transform(block.Txns, func(txn *MsgDeSoTxn) *BlockHash {
+		return txn.Hash()
+	})
+	// Connect the block to the view!
+	utxoOpsForBlock, err := utxoView.ConnectBlock(block, txHashes, true /*verifySignatures*/, bc.eventManager, block.Header.Height)
+	if err != nil {
+		// TODO: rule error handling? mark blocks invalid?
+		return errors.Wrapf(err, "commitBlockPoS: Problem connecting block to view: ")
+	}
+	// Put the block in the db
+	// Note: we're skipping postgres.
+	// TODO: this is copy pasta from ProcessBlockPoW. Refactor.
+	blockNode.Status |= StatusBlockCommitted
+	err = bc.db.Update(func(txn *badger.Txn) error {
+		if bc.snapshot != nil {
+			bc.snapshot.PrepareAncestralRecordsFlush()
+			defer bc.snapshot.StartAncestralRecordsFlush(true)
+			glog.V(2).Infof("commitBlockPoS: Preparing snapshot flush")
+		}
+		// Store the new block in the db under the
+		//   <blockHash> -> <serialized block>
+		// index.
+		// TODO: In the archival mode, we'll be setting ancestral entries for the block reward. Note that it is
+		// 	set in PutBlockWithTxn. Block rewards are part of the state, and they should be identical to the ones
+		// 	we've fetched during Hypersync. Is there an edge-case where for some reason they're not identical? Or
+		// 	somehow ancestral records get corrupted?
+		if innerErr := PutBlockWithTxn(txn, bc.snapshot, block); innerErr != nil {
+			return errors.Wrapf(innerErr, "commitBlockPoS: Problem calling PutBlock")
+		}
+
+		// Store the new block's node in our node index in the db under the
+		//   <height uin32, blockHash BlockHash> -> <node info>
+		// index.
+		if innerErr := PutHeightHashToNodeInfoWithTxn(txn, bc.snapshot, blockNode, false /*bitcoinNodes*/); innerErr != nil {
+			return errors.Wrapf(innerErr, "commitBlockPoS: Problem calling PutHeightHashToNodeInfo before validation")
+		}
+
+		// Set the best node hash to this one. Note the header chain should already
+		// be fully aware of this block so we shouldn't update it here.
+		if innerErr := PutBestHashWithTxn(txn, bc.snapshot, blockNode.Hash, ChainTypeDeSoBlock); innerErr != nil {
+			return errors.Wrapf(innerErr, "commitBlockPoS: Problem calling PutBestHash after validation")
+		}
+		// Write the utxo operations for this block to the db so we can have the
+		// ability to roll it back in the future.
+		if innerErr := PutUtxoOperationsForBlockWithTxn(txn, bc.snapshot, uint64(blockNode.Height), blockNode.Hash, utxoOpsForBlock); innerErr != nil {
+			return errors.Wrapf(innerErr, "ProcessBlock: Problem writing utxo operations to db on simple add to tip")
+		}
+		return nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "commitBlockPoS: Problem putting block in db: ")
+	}
+	if err = utxoView.FlushToDb(uint64(blockNode.Height)); err != nil {
```

Comment:
Hmm I think this has to be in the previous badger txn and use utxoView.FlushToDbWithTxn(). If you don't do this, you run the risk of the db getting corrupted during a restart. It would also be a reeeally hard to find bug...

Reply:
Put another way, everything we write to the DB when we commit a block has to be atomic. The only way to do that is to commit it all with the same badger txn.

Reply:
So the REALLY annoying thing is that `utxoView.FlushToDB` calls `PrepareAncestralRecordsFlush`, and having nested calls to `PrepareAncestralRecordsFlush` raises an error. So do I not call the prepare ancestral records flush stuff at the top here? or how should it work? Trying to do stuff similar to what's in ProcessBlockPoW

Reply:
discussed offline - will write a flush to db func that doesn't call PrepareAncestralRecords flush itself.

Reply:
resolved in #800